cmd/tools/generate/counter.go

```diff
@@ -1825,3 +1877,70 @@ func getAllExportedLabels(t *node.Node, counterContents []string) ([]string, []s
 	}
 	return metricLabels, append(labels, metricLabels...), isInstanceLabels
 }
+
+func visitDashboard(dirs []string, metricsPanelMap map[string]PanelData, eachDash func(data []byte, metricsPanelMap map[string]PanelData)) {
+	for _, dir := range dirs {
+		err := filepath.Walk(dir, func(path string, _ os.FileInfo, err error) error {
+			if strings.Contains(path, "influxdb") {
+				return nil
+			}
+			if err != nil {
+				log.Fatal("failed to read directory:", err)
+			}
+			ext := filepath.Ext(path)
+			if ext != ".json" {
+				return nil
+			}
+			data, err := os.ReadFile(path)
+			if err != nil {
+				log.Fatalf("failed to read dashboards path=%s err=%v", path, err)
+			}
+			eachDash(data, metricsPanelMap)
+			return nil
+		})
+		if err != nil {
+			log.Fatal("failed to read dashboards:", err)
+		}
+	}
+}
+
+func visitExpressions(data []byte, metricsPanelMap map[string]PanelData) {
+	// collect all expressions
+	expressions := make([]grafana.ExprP, 0)
+	dashboard := gjson.GetBytes(data, "title").String()
+	uid := gjson.GetBytes(data, "uid").String()
+	link := "d/" + uid + "/" + strings.ToLower(strings.Replace(dashboard, ": ", "3a-", 1)) + "?orgId=1&viewPanel="
+	gjson.GetBytes(data, "panels").ForEach(func(key, value gjson.Result) bool {
+		grafana.DoTarget("", "", key, value, func(path string, expr string, kind string, id string, title string, rowTitle string) {
+			expressions = append(expressions, grafana.NewExpr(path, expr, kind, id, title, rowTitle))
+		})
+		tp := value.Get("type").String()
+		rowTitle := ""
+		if tp == "row" {
+			rowTitle = value.Get("title").String()
+		}
+		value.Get("panels").ForEach(func(key2, value2 gjson.Result) bool {
+			pathPrefix := fmt.Sprintf("panels[%d].", key.Int())
+			grafana.DoTarget(pathPrefix, rowTitle, key2, value2, func(path string, expr string, kind string, id string, title string, rowTitle string) {
+				expressions = append(expressions, grafana.NewExpr(path, expr, kind, id, title, rowTitle))
+			})
+			return true
+		})
+		return true
+	})
+
+	for _, expr := range expressions {
+		allMatches := metricRe.FindAllStringSubmatch(expr.Expr, -1)
+		for _, match := range allMatches {
+			m := match[1]
+			if m == "" {
+				continue
+			}
+
+			key := dashboard + expr.RowTitle + expr.Kind + expr.PanelTitle + expr.PanelID
+			if !slices.Contains(metricsPanelMap[m].Keys, key) {
```

Comment:
Instead of searching over a slice, would it make more sense to store these in a map so we can do O(1) lookups?